# Ejercicio 1

##### Coordenadas polares y cartesianas

Considerando que la transformación de coordenadas polares a cartesianas se realiza a partir de las siguientes expresiones: 

Cartesianas a polares:

$$ r = \sqrt{x^2 + y^2}$$

$$ \theta = arctan(y/x)$$

Polares a cartesianas:

$$ x = rsin(\theta)$$

$$y = rcos(\theta)$$

Por lo que el código para realizar dichas transformaciones será respectivamente:

In [12]:
import math as mt
print("Ingresa tus coordenadas del punto (r,theta), con theta en grados:")
r = float(input())
theta = float(input())
theta_rad = mt.radians(theta)

xp = r * mt.cos(theta_rad)
yp = r * mt.sin(theta_rad)

print("La tranformación de tu punto (" + str(r) + "," + str(theta) + ") en coord. polares a cartesianas es:")
print("(" + str(xp) + "," + str(yp) + ")")

Ingresa tus coordenadas del punto (r,theta), con theta en grados:


 1
 90


La tranformación de tu punto (1.0,90.0) en coord. polares a cartesianas es:
(6.123233995736766e-17,1.0)


In [2]:
import math as mt
print("Ingresa tus coordenadas del punto (x,y):")
x = float(input())
y = float(input())

r_c = mt.sqrt(x**2 + y**2)
theta_c = mt.atan2(y,x)

print("La tranformación de tu punto (" + str(x) + "," + str(y) + ") de coord. cartesianas a polares es:")
print("(" + str(r_c) + "," + str(theta_c) + ")")

Ingresa tus coordenadas del punto (x,y):


 1
 1


La tranformación de tu punto (1.0,1.0) de coord. cartesianas a polares es:
(1.4142135623730951,0.7853981633974483)


# Ejercicio 2

##### Productos vectoriales y matriciales

Escribe el pseudocódigo de los algoritmos que realicen las siguientes operaciones vectoriales; así como los respectivos programas en python, sin usar bibliotecas como numpy ni cualquier otra (i.e. usando sólo listas) que implementen dichos algoritmos:

##### Producto interior en $\mathbb{R}^3$

En este caso, para la implementación el usuario del programa debe que insertar las entradas de los vectores. 

Definimos dos vectores:

$\vec{v}$ = ($v_1$, $v_2$, $v_3$)

$\vec{u}$ = ($u_1$, $u_2$, $u_3$)


Defnimos un contador *n* para que vaya sumando cada uno de los elementos:

n = 0

Entramos en un ciclo for, contando el número de entradas  del vector a multiplicar:

**for** i de 1 ⮕ 3 : 

    n = n + v_i * u_i

Imprimir *n*. Debido a que es la suma de cada producto entrada por entrada.

In [48]:
v = []
u = []

print("Ingresa los valores de tu vector v:")
for i in range(3):
    M = float(input())
    v.append(M) #agregamos las entradas al vector vacío v

print("Ingresa los valores de tu vector u:")
for i in range(3):
    K = float(input())
    u.append(K) #agregamos las entradas al vector vacío u

f = 0
for i in range(3):
    f = f + v[i]*u[i]
print(f)

Ingresa los valores de tu vector v:


 7
 8
 9


Ingresa los valores de tu vector u:


 6
 2.2
 5.99


113.51


##### Producto interior de dos vectores en $\mathbb{R}^n$

Para la implementación, en este caso, encuentra una manera de generar dos vectores diferentes en $\mathbb{R}^n$

Definimos dos vectores:

$\vec{v}$ = ($v_1$, $v_2$, ..., $v_n$)

$\vec{u}$ = ($u_1$, $u_2$, ..., $u_n$)


Defnimos un contador *n* para que vaya sumando cada uno de los elementos:

n = 0

Entramos en un ciclo for, contando el número de entradas  del vector a multiplicar:

**for** i de 1 ⮕ n : 

    n = n + v_i * u_i

Imprimir *n*. Debido a que es la suma de cada producto entrada por entrada.

In [52]:
import numpy as np
dim = int(input("Ingresa la dimensión de tus vectores:"))
m = dim

v = np.random.randint(1,100, size = m)
u = np.random.randint(1,100, size = m)

f = 0
for i in range(dim):
    f = f + v[i]*u[i]
print("Siendo tus vectores: v = ", v, "y u = ", u, ". El producto punto entre ellos es igual a ", f) 

Ingresa la dimensión de tus vectores: 3


Siendo tus vectores: v =  [79 23 87] y u =  [80 83  1] . El producto punto entre ellos es igual a  8316


##### Producto de una matriz en $\mathbb{A} \in \mathbb{R}^{nxn}$ por un vector de $\mathbb{R}^n$

Para la implementación las entradas tanto del vector como de la matriz deben leerse de un archivo de datos (que tú puedes generar).

Generamos un vector y una matriz: 

$v = (v_1, v_2, ..., v_n)$ 

$M = ((m_{11}, m_{12}, ..., m_{1n}), ..., (m_{n1}, m_{n2}, ..., m_{nn}))$


Donde $M$ lo definimos como una lista de listas para visualizarlo mejor a la hora de implementarlo en el código. De esta manera, siendo que el producto de una matriz por un vector genera un vector, definimos un vector vacío para ir guardando las entradas resultantes que vayamos generando:

$u = (u_1, u_2, ..., u_n)$

Entramos en un ciclo **for** para poder ir recorriendo cada una de las entradas: 

Para cada i de 1 ⮕ n: 

    Para cada j de 1 ⮕ n:

        u_i = u_i + A_ij * v_i

        Añadir elemento u_i a la entrada i del vector u

Imprimir el vector u

##### Producto de dos matrices $\mathbb{A}, \mathbb{A} \in \mathbb{R}^{nxn}$

En la implementación de las entradas de ambas matrices deben leerse de un archivo de datos (que tú generes).

Generamos un vector y una matriz: 

$N = ((n_{11}, n_{12}, ..., n_{1n}), ..., (n_{n1}, n_{n2}, ..., n_{nn}))$ 

$M = ((m_{11}, m_{12}, ..., m_{1n}), ..., (m_{n1}, m_{n2}, ..., m_{nn}))$

Donde $N$ y $M$ las definimos como una lista de listas para visualizarlo mejor a la hora de implementarlo en el código. De esta manera, siendo que el producto de una matriz por otra matriz de misma dimensión genera otra matriz de dicha dim., definimos una matriz vacía para ir guardando las entradas resultantes que vayamos generando:

$L = ((l_{11}, l_{12}, ..., l_{1n}), ..., (l_{n1}, l_{n2}, ..., l_{nn}))$

Entramos en dos ciclos **for** para poder ir recorriendo cada una de las entradas: 

Para cada i de 1 ⮕ n: 

    Para cada j de 1 ⮕ n:

        l_ij = l_i + n_ij * m_ji

        Añadir elemento u_ij a la entrada ij de la matriz L

Imprimir la matriz L

##### Funciones

Convierte tus códigos en funciones y generaliza todo en una única función que realice el producto de dos matrices $\mathbb{A} \in \mathbb{R}^{nxn}$ y $\mathbb{B} \in \mathbb{R}^{nxn}$, así como el cálculo de cualquiera de los incisos anteriores, dependiendo de la entrada que se le da a la función. En todos los casos, las entradas deben leerse de un archivo de datos. 

El pseudocódigo como se deduce de los anteriores ejemplos se obtiene particularmente del productor punto entre los vectores de dimensión n. Esto debido a que, particularmente en el caso de las matrices, estas son corformadas por listas en una lista, y dichas listas se pueden ver como entradas de los vectores:

**definimos** producto(v,u,m):

    v = [v_1, v_2, ..., v_n]
    
    u = [u_1, u_2, ..., u_n]
    
    n = 0

    for i in range(m):
        n = n + v_i * v_i

    Imprimir n

# Ejercicio 3

##### Caida de una pelota desde una torre

Se deja caer una pelota desde una torre de altura $h$. Tienen velocidad inicial cero y acelera hacia abajo con la gravedad. Escribe un programa que le pida al usuario que ingrese la altura en metros de la torre y luego calcule e imprima el tiempo en segundos hasta que la pelota toque el suelo (ignora la resistencia del aire).

Siendo que se trata de un movimiento acelerado, podemos aplicar las ecuaciones para obtener el tiempo del choque de la pelota contra el suelo. Para ello, siendo que: 

$$ V_{f} = V_{i} + at$$
$$ V_{f}^2 = V_{i}^2 + 2ah$$

Donde $V_{f}$ y $V_{i}$ son las velocidades final e inicial, $a$ es la aceleración, en este caso, tomará el valor de $g=9.8 m/s$ y $h$ es la altura. Sin embargo, al ser que $V_{i}=0$, entonces podemos obtener el tiempo con:  

$$t = \frac{\sqrt{2ah}}{a}$$

In [13]:
import math as mt
a = 9.8
h = int(input("Ingresa la altura desde la que se deja caer la pelota:"))

t = (mt.sqrt(vi**2 + 2*a*h)) / a

print("La pelota toca el suelo a un tiempo de", t, "s")

Ingresa la altura desde la que se deja caer la pelota: 100


La pelota toca el suelo a un tiempo de 4.5175395145262565 s


Después usa tu programa para calcular el tiempo de una pelota lanzada desde una torre de 100m de altura.

Al calcular, se obtiene que a una altura de $100m$, tarda un tiempo de $4.5175395145262565 s$ para que la pelota choque con el suelo.

# Ejercicio 4

Se va a lanzar un satélite en una órbita circular alrededor de la Tierra de modo que orbite el planeta una vez cada $T$ segundos. 

a) Demuestre que la altitud $h$ sobre la superficie de la Tierra que debe tener el satélite es 
$$ h = (\frac{GMT^2}{4 \pi^2})^{1/3} - R $$

*Ver el archivo anexado en Classroom

b) Escribe un programa que le pida al usuario que ingrese el valor deseado de T y luego calcula e imprima la altitud correcta en metros.

Siendo que la altura debe que ser dada en metros, basta con considerar el radio de la tierra como $R=6371000 m$,
de esta manera, utilizando la ecuación del inciso a), se tiene que:

In [34]:
import math as mt
def altitud(T): # Definimos una función para calcular la altitud
    G = 6.67E-11
    M = 5.97E24
    R = 6371000
    h = ((G*M*(T**2))/(4*(mt.pi**2)))**(1/3)-R
    return h
    
Tp = input("¿Deseas ingresar el período en horas (h) o minutos (m)?")

if Tp == "h": #usamos el if para evitar pedir al usuario que ponga sólo T en minutos
    T = 60*60*int(input("Ingresa el período T al que orbita el satélite:"))
    H = altitud(T)
else: 
    T = 60*int(input("Ingresa el período T al que orbita el satélite:"))
    H = altitud(T)

print("El satélite órbita a una altura de", H, "m respecto de la Tierra") 

¿Deseas ingresar el período en horas (h) o minutos (m)? m
Ingresa el período T al que orbita el satélite: 45


El satélite órbita a una altura de -2181559.8978108233 m respecto de la Tierra


c) Utiliza tu programa para calular las altitudes de los satélites que orbitan la Tierra una vez al día (la llamada órbita geoestacionaria), una vez cada 90 minutos y una vez cada 45 minutos. ¿Qué concluyes de este último cálculo?

Para la órbita geoestacionaria, la altitud del satélite es de $35855910.176174976 m$; para el satélite con período de $90 min$ es de $279321.6253728606m$ y por último, para el satélite de $45 min$ es de $-2181559.8978108233 m$. El que la altura para el último satélite sea negativa implica que el radio del planeta en que órbite el satélite debe que ser menor al de la Tierra. Esto es debido a que T debería que ser mayor a $5063.38783$. Esto se obtiene a partir de que:

$$h > 0 $$
$$ (\frac{GMT^2}{4 \pi^2})^{1/3} - R > 0 $$
$$ \frac{GMT^2}{4 \pi^2} > R^3 $$
$$ T^2 > \frac{4 \pi R^3}{GM} $$
$$ T > 2 \pi (\frac{R^3}{GM})^{1/2} \approx 5063.38783 s = 84.38 min $$

# Ejercicio 5

Una nave espacial viaja desde la Tierra en línea recta a una velocidad relativista $v$ a otro planeta a $x$ años luz de distancia. Escribe un programa que le pida al usuario el valor de $x$ y la velocidad $v$ como una fracción de la velocidad de la luz $c$, y que imprima el tiempo en años que tarda la nave espacial en llegar a su destino. 

a) en el marco de reposo de un observador en la Tierra y

Siendo que para el observador de la Tierra, el tiempo viene dado por 
$$ t_{T} = \frac{x}{v}$$
Se tiene que el código va a ser:

In [39]:
x = int(input("Ingresa la distancia (en años luz) que recorrerá la nave:"))
v = float(input("Ingresa la velocidad de la nave (fracción de años luz):"))

t_T = x / v

print("El tiempo que tarda la nave en llegar a su destino visto desde la Tierra es de", t_T, "años luz.")

Ingresa la distancia (en años luz) que recorrerá la nave: 10
Ingresa la velocidad de la nave (fracción de años luz): 0.99


El tiempo que tarda la nave en llegar a su destino visto desde la Tierra es de 10.1010101010101 años


b) como lo percibiría un pasajero a bordo de la nave.

Siendo que el pasajero a bordo de la nave está sujeto a la dilatación del tiempo que experimenta debido a la velocidad con la que va viajando, éste se describe a partir de la transformación de Lorentz:
$$t_{nave} = \frac{x}{v}\sqrt{1-\frac{v}{c}^2}$$
Con $c=1$ para normalizar los resultados. Calculando el tiempo transcurrido para el pasajero a bordo, se tienen que el código será:

In [40]:
import math as mt
x = int(input("Ingresa la distancia (en años luz) que recorrerá la nave:"))
v = float(input("Ingresa la velocidad de la nave (fracción de años luz):"))

t_n = (x / v) * mt.sqrt(1-v**2)

print("El tiempo que tarda la nave en llegar a su destino percibido desde la nave es de", t_n, "años luz.")

Ingresa la distancia (en años luz) que recorrerá la nave: 10
Ingresa la velocidad de la nave (fracción de años luz): 0.99


El tiempo que tarda la nave en llegar a su destino percibido desde la nave es de 1.424922826228878 años luz


Usa tu programa para calcular las respuestas para un planeta a 10 años luz de distancia con $v = 0.99c$

Para la persona en la Tierra, se tiene que el tiempo de $10.1010101010101$ años luz. Mientras que para el pasajero en la nave, será de $1.424922826228878$ años luz.